File: occasum-json/src/main/java/pw/stamina/occasum/dao/json/JsonLoadingPropertyNodeVisitor.java
package pw.stamina.occasum.dao.json;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import pw.stamina.occasum.node.PropertyNode;
import pw.stamina.occasum.node.visit.AbstractPropertyNodeVisitor;
import pw.stamina.occasum.node.visit.PropertyNodeVisitor;
import pw.stamina.occasum.properties.PropertyParseException;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

final class JsonLoadingPropertyNodeVisitor extends AbstractPropertyNodeVisitor {
    private static final String JSON_OBJECT_PROPERTY_VALUE_NAME = "value";

    private final JsonElement element;
    private final List<Exception> exceptions;

    JsonLoadingPropertyNodeVisitor(JsonElement element) {
        this(element, new ArrayList<>());
    }

    private JsonLoadingPropertyNodeVisitor(JsonElement element,
                                           List<Exception> exceptions) {
        this.element = element;
        this.exceptions = exceptions;
    }

    @Override
    public void visitNode(PropertyNode node) {
        if (!node.hasProperty()) {
            return;
        }

        findPropertyValueFromElement().ifPresent(value -> {
            try {
                node.getProperty().parseAndSet(value);
            } catch (PropertyParseException e) {
                exceptions.add(e);
            }
        });
    }

    private Optional<String> findPropertyValueFromElement() {
        if (element.isJsonPrimitive()) {
            return Optional.of(element.getAsString());
        }

        if (element.isJsonObject()) {
            JsonObject object = element.getAsJsonObject();
            String name = JSON_OBJECT_PROPERTY_VALUE_NAME;

            if (object.has(name)) {
                JsonPrimitive value = object.getAsJsonPrimitive(name);
                return Optional.of(value.getAsString());
            }
        }

        return Optional.empty();
    }

    @Override
    protected PropertyNodeVisitor createChildVisitor(PropertyNode childNode) {
        if (!element.isJsonObject()) {
            return null;
        }

        JsonObject object = element.getAsJsonObject();
        String name = childNode.getId();

        if (!object.has(name)) {
            return null;
        }

        JsonElement childElement = object.get(name);
        return new JsonLoadingPropertyNodeVisitor(childElement, exceptions);
    }

    public List<Exception> getExceptions() {
        return exceptions;
    }
}

File: occasum-json/src/main/kotlin/pw/stamina/occasum/dao/json/JsonLoadingPropertyNodeVisitor.kt
package pw.stamina.occasum.dao.json

import com.google.gson.JsonElement
import pw.stamina.occasum.dao.PropertyDao
import pw.stamina.occasum.node.PropertyNode
import pw.stamina.occasum.node.visit.PropertyNodeVisitor
import pw.stamina.occasum.properties.PropertyParseException
import java.util.*

internal class JsonLoadingPropertyNodeVisitor private constructor(
        private val element: JsonElement,
        private val exceptions: MutableList<Exception>
) : PropertyNodeVisitor {

    constructor(element: JsonElement) : this(element, ArrayList<Exception>())

    override fun visitNode(node: PropertyNode) {
        val property = node.property ?: return

        findPropertyValueFromElement().ifPresent { value ->
            try {
                property.parseAndSet(value)
            } catch (e: PropertyParseException) {
                exceptions.add(e)
            }
        }
    }

    private fun findPropertyValueFromElement(): Optional<String> {
        if (element.isJsonPrimitive) {
            return Optional.of(element.asString)
        }

        if (element.isJsonObject) {
            val `object` = element.asJsonObject
            val name = PropertyDao.RESERVED_SERIALIZED_VALUE_NAME

            if (`object`.has(name)) {
                val value = `object`.getAsJsonPrimitive(name)
                return Optional.of(value.asString)
            }
        }

        return Optional.empty()
    }

    override fun visitChildNode(childNode: PropertyNode): PropertyNodeVisitor? {
        if (!element.isJsonObject) {
            return null
        }

        val `object` = element.asJsonObject
        val name = childNode.id

        if (!`object`.has(name)) {
            return null
        }

        val childElement = `object`.get(name)
        return JsonLoadingPropertyNodeVisitor(childElement, exceptions)
    }

    fun getExceptions(): List<Exception> {
        return exceptions
    }
}
